File: 021.MergeTwoSortedLists/wyrj/MergeTwoSortedList.c
/* 4ms */
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* mergeTwoLists(struct ListNode* l1, struct ListNode* l2) {
    struct ListNode *prev = NULL;
    struct ListNode *head = NULL;
    struct ListNode *node1 = l1, *node2 = l2;
    while(NULL != node1 && NULL != node2) {
        if (node1->val <= node2->val) {
            if (NULL == prev) {
                prev = node1;
                head = node1;
            } else {
                prev->next = node1;
                prev = prev->next;
            }
            node1 = node1->next;
        } else {
            if (NULL == prev) {
                prev = node2;
                head = node2;
            } else {
                prev->next = node2;
                prev = prev->next;
            }
            node2 = node2->next;
        }
    }
    if (NULL == prev) {
        head = (NULL != node1) ? node1 : node2;
    } else {
        prev->next = (NULL != node1) ? node1 : node2;
    }
    return head;
}

File: 021.MergeTwoSortedLists/wyrj/MergeTwoSortedList.js
/* 152ms */
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var mergeTwoLists = function(l1, l2) {
    var prev = null, head = null, node1 = l1, node2 = l2;
    while(null !== node1 && null !== node2) {
        if (node1.val <= node2.val) {
            if (null === prev) {
                head = node1;
                prev = node1;
            } else {
                prev.next = node1;
                prev = prev.next;
            }
            node1 = node1.next;
        }
        else {
            if (null === prev) {
                head = node2;
                prev = node2;
            } else {
                prev.next = node2;
                prev = prev.next;
            }
            node2 = node2.next;
        }
    }
    if (null !== node1) {
        if (null === prev) {
            head = node1;
        } else {
            prev.next = node1;
        }
    } else {
        if (null === prev) {
            head = node2;
        } else {
            prev.next = node2;
        }
    }
    return head;
};